File: src/main/java/xyz/phanta/cbtweaker/template/TemplateRegistry.java
package xyz.phanta.cbtweaker.template;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import xyz.phanta.cbtweaker.CbtMod;
import xyz.phanta.cbtweaker.util.ConfigException;
import xyz.phanta.cbtweaker.util.helper.DataLoadUtils;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.Map;


public class TemplateRegistry<T> {

    private final String templateTypeName;
    private final Path templateFile;
    private final Resolver<T> resolver;
    private final Map<String, T> templateTable = new HashMap<>();

    public TemplateRegistry(String templateTypeName, Path templateFile, Resolver<T> resolver) {
        this.templateTypeName = templateTypeName;
        this.templateFile = templateFile;
        this.resolver = resolver;
    }

    public void loadTemplates() {
        try {
            if (!Files.isRegularFile(templateFile)) {
                return;
            }
            JsonObject templatesDto = DataLoadUtils.readJsonFile(templateFile).getAsJsonObject();
            for (Map.Entry<String, JsonElement> templateEntry : templatesDto.entrySet()) {
                try {
                    templateTable.put(templateEntry.getKey(), resolver.resolve(templateEntry.getValue()));
                } catch (ConfigException e) {
                    CbtMod.LOGGER.warn("Ignoring bad " + templateTypeName + " template: " + templateEntry.getKey(), e);
                }
            }
        } catch (Exception e) {
            throw new IllegalArgumentException("Failed to load " + templateTypeName + " templates!", e);
        }
    }

    public T resolve(JsonElement dto) {
        if (dto.isJsonPrimitive() && dto.getAsJsonPrimitive().isString()) {
            String s = dto.getAsString();
            if (!s.isEmpty() && s.charAt(0) == '#') {
                T template = templateTable.get(s.substring(1));
                if (template == null) {
                    throw new ConfigException("Unknown " + templateTypeName + " template: " + s);
                }
                return template;
            }
        }
        return resolver.resolve(dto);
    }

    public interface Resolver<T> {

        T resolve(JsonElement dto);

    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/template/TemplateRegistry.kt
package st.evening.mc.cbtweaker.template

import st.evening.mc.cbtweaker.CbTweaker
import st.evening.mc.prelude.api.data.ser.SerializationException
import st.evening.mc.prelude.api.data.tjson.JsonPath
import st.evening.mc.prelude.api.data.tjson.TJson
import st.evening.mc.prelude.api.data.tjson.TypedJsonParser
import st.evening.mc.prelude.api.data.tjson.forEachValue
import java.nio.file.Files
import java.nio.file.Path
import kotlin.io.path.readText

class TemplateRegistry<T>(
    private val templateTypeName: String,
    private val templateFile: Path,
    private val resolver: context(JsonPath) (TJson) -> T
) {
    private val templateTable: MutableMap<String, T> = mutableMapOf()

    fun loadTemplates() {
        try {
            if (!Files.isRegularFile(templateFile)) return
            JsonPath.atRoot {
                TypedJsonParser.parseObject(templateFile.readText()).forEachValue { key, dto ->
                    try {
                        register(key, resolver(dto))
                    } catch (e: SerializationException) {
                        CbTweaker.logger.warn("Ignoring bad {} template: {}", templateTypeName, key, e)
                    }
                }
            }
        } catch (e: Exception) {
            throw IllegalStateException("Failed to load $templateTypeName templates!", e)
        }
    }

    fun register(key: String, template: T) {
        if (templateTable.put(key, template) != null) {
            throw IllegalStateException("Duplicate $templateTypeName template: $key")
        }
    }

    context(_: JsonPath)
    fun resolve(dto: TJson): T {
        if (dto is TJson.String) {
            val value = dto.value
            if (value.isNotEmpty() && value[0] == '#') {
                val key = value.substring(1)
                return templateTable[key]
                    ?: throw SerializationException.withPath("Unknown $templateTypeName template: $key")
            }
        }
        return resolver(dto)
    }
}
